addons/mail/models/ir_attachment.py

```diff
@@ -34,3 +34,14 @@ def register_as_main_attachment(self, force=True):
                     related_record.message_main_attachment_id = self
                 except AccessError:
                     pass
+
+    def unlink(self):
+        """
+        Delete all messages without text content and linked to this attachment.
+        """
+        empty_messages = self.env['mail.message'].search([
+            ('attachment_ids', 'IN', self.ids),
+            ('body', '=', '')
+        ])
+        empty_messages.unlink()
+        return super().unlink()
```

Comment:
I think the concept is good.

Here are what I feel are the 5 main points of attention in regard to this change:

1. (covering too much) Ensure this properly covers all cases and doesn't delete too many messages. What else could make a message considered not empty? Tracking values? Specific subtypes? If the message is the first message of the thread and has `child_ids`, this may be problematic also.

2. (not covering enough) Is empty string body the only possibility for a message to be empty? Isn't it a variation of `</p>` as body also possible? Or could body be `null`?

3. (performance) Ensure this is not a disaster in regard to performance on a database with millions of message. This will be called whenever any attachment is deleted, whether it is linked to a message or not. This will include assets when they are generated, documents, etc. We can populate data to test this. We can also always ask someone with prod access to run the query on it and see how it goes.

4. (rights) Will the person allowed to delete the attachment be also allowed to delete the message? Shouldn't we do some strange access right check before calling the method on attachment, and other strange check before deleting the related messages?

5. (propagate info) Should we notify the JS with the bus?


Also more a general concern, we currently don't support deleting messages from users... For some reasons. Doing this will probably overlap with some of those reasons, and might not be doable at all in current state? Or if it is, we could be ready to support deleting messages?

I don't remember, have you checked how it worked in 13.3 for comparison?

Reply:
in 13.3, it's done like it was in this PR before : just hide empty messages

Reply:
> I think the concept is good.
> 
> Here are what I feel are the 5 main points of attention in regard to this change:
> 
> 1. (covering too much) Ensure this properly covers all cases and doesn't delete too many messages. What else could make a message considered not empty? Tracking values? Specific subtypes? If the message is the first message of the thread and has `child_ids`, this may be problematic also.

In 13.3, here is the code used to hide messages :  

```js
    /**
     * State whether this message is empty
     *
     * @return {boolean}
     */
    isEmpty: function () {
        return !this.hasTrackingValues() &&
        !this.hasAttachments() &&
        !this.getBody();
    },
```
So I suppose only tracking values are missing here.

> 2. (not covering enough) Is empty string body the only possibility for a message to be empty? Isn't it a variation of `</p>` as body also possible? Or could body be `null`?

I test multiple things and it appears that it's always saved as `''` in database, I suppose `fields.Html` by calling `html_sanitize` removes useless tags.

> 3. (performance) Ensure this is not a disaster in regard to performance on a database with millions of message. This will be called whenever any attachment is deleted, whether it is linked to a message or not. This will include assets when they are regenerated, deletion of documents, garbage collect of pending attachments, etc. We can populate data to test this. We can also always ask someone with prod access to run the query on it and see how it goes.

Indeed this should probably be tested.

> 4. (rights) Will the person allowed to delete the attachment be also allowed to delete the message? Shouldn't we do some strange access right check before calling the method on attachment, and other strange check before deleting the related messages?

As you said, message can be deleted as there is no lost data, so I'm wondering if it's really needed to check rights for that ? If the user is able to delete a given attachment, making a message empty, then the message could be deleted like part of a garbage collecting process.

> 5. (propagate info) Should we notify the JS with the bus?
> 
> Also more a general concern, we currently don't support deleting messages from users... For some reasons. Doing this will probably overlap with some of those reasons, and might not be doable at all in current state? Or if it is, we could be ready to support deleting messages?
> 

It probably should so that all users are aware of a deleted message but what about performance again ?

> I don't remember, have you checked how it worked in 13.3 for comparison?

in 13.3, it's done like it was in this PR before : just hide empty messages